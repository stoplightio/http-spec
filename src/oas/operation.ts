import { isPlainObject } from '@stoplight/json';
import type { DeepPartial, IHttpEndpointOperation } from '@stoplight/types';
import type { OpenAPIObject, OperationObject, PathsObject } from 'openapi3-ts';
import type { Spec } from 'swagger-schema-official';
import pickBy = require('lodash.pickby');

import { isBoolean, isString } from '../guards';
import type { EndpointOperationConfig, Fragment, HttpEndpointOperationTransformer } from '../types';
import { TransformerContext, TranslateFunction } from '../types';
import { extractId } from '../utils';
import { getExtensions } from './accessors';
import { toExternalDocs } from './externalDocs';
import { translateToTags } from './tags';
import { translateToSecurityDeclarationType } from './transformers';

const DEFAULT_METHODS = ['get', 'post', 'put', 'delete', 'options', 'head', 'patch', 'trace'];

export const OPERATION_CONFIG: EndpointOperationConfig = {
  type: 'operation',
  documentProp: 'paths',
  nameProp: 'path',
};

export const WEBHOOK_CONFIG: EndpointOperationConfig = {
  type: 'webhook',
  documentProp: 'webhooks',
  nameProp: 'name',
};

export function transformOasEndpointOperations<
  T extends Fragment & DeepPartial<Spec | OpenAPIObject>,
  TEndpoint extends IHttpEndpointOperation,
>(
  document: T,
  transformer: HttpEndpointOperationTransformer<any, TEndpoint>,
  config: EndpointOperationConfig,
  methods: string[] | null = DEFAULT_METHODS,

  ctx?: TransformerContext<T>,
): TEndpoint[] {
  const entries = isPlainObject(document[config.documentProp]) ? Object.entries(document[config.documentProp]) : [];

  return entries.flatMap(([name, value]) => {
    if (!isPlainObject(value)) return [];

    let operations = Object.keys(value);
    if (methods !== null) {
      operations = operations.filter(pathKey => methods.includes(pathKey));
    }

    return operations.map(method =>
      transformer({
        document,
        name,
        method,
        config,
        ctx,
      }),
    );
  });
}

export const transformOasEndpointOperation: TranslateFunction<
  DeepPartial<OpenAPIObject> | DeepPartial<Spec>,
  [config: EndpointOperationConfig, name: string, method: string],
  Omit<IHttpEndpointOperation, 'responses' | 'request' | 'servers' | 'security' | 'callbacks'>
> = function ({ type, documentProp, nameProp }: EndpointOperationConfig, name: string, method: string) {
  const pathObj = this.maybeResolveLocalRef(this.document?.[documentProp]?.[name]) as PathsObject;
  if (typeof pathObj !== 'object' || pathObj === null) {
    throw new Error(`Could not find ${[documentProp, name].join('/')} in the provided spec.`);
  }

  const obj = this.maybeResolveLocalRef(pathObj[method]) as OperationObject;
  if (!obj) {
    throw new Error(`Could not find ${[documentProp, name, method].join('/')} in the provided spec.`);
  }

  const serviceId = (this.ids.service = String(this.document['x-stoplight']?.id));
  if (type === 'operation') {
    this.ids.path = this.generateId.httpPath({ parentId: serviceId, path: name });
  } else {
    this.ids.webhookName = this.generateId.httpWebhookName({ parentId: serviceId, name });
  }
  let id: string;
  if (this.context === 'callback') {
    id = this.ids.operation =
      extractId(obj) ??
      this.generateId.httpCallbackOperation({
        parentId: serviceId,
        method,
        path: name,
      });
  } else if (type === 'operation') {
    id = this.ids.operation =
      extractId(obj) ?? this.generateId.httpOperation({ parentId: serviceId, method, path: name });
  } else {
    id = this.ids.webhook = extractId(obj) ?? this.generateId.httpWebhook({ parentId: serviceId, method, name });
  }

  this.parentId = id;

  this.context = type;

  return {
    id,

    method,
    [nameProp]: name,

    tags: translateToTags.call(this, obj.tags),
    extensions: getExtensions(obj),

    ...pickBy(
      {
        deprecated: obj.deprecated,
        internal: obj['x-internal'],
      },
      isBoolean,
    ),

    ...pickBy(
      {
        iid: obj.operationId,
        description: obj.description,
        summary: obj.summary,
      },
      isString,
    ),

    securityDeclarationType: translateToSecurityDeclarationType(obj),
    ...toExternalDocs(obj.externalDocs),
  };
};
