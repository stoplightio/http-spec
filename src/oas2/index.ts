export * from './operation';
export * from './service';
export * from './transformers/params';
