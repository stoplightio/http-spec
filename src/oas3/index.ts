export * from './operation';
export * from './service';
