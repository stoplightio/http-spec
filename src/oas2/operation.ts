import type { DeepPartial, IHttpOperation } from '@stoplight/types';
import type { Spec } from 'swagger-schema-official';

import { createContext } from '../oas/context';
import { OPERATION_CONFIG, transformOasEndpointOperation, transformOasEndpointOperations } from '../oas/operation';
import { Oas2HttpOperationTransformer } from '../oas/types';
import type { Fragment } from '../types';
import { TransformerContext } from '../types';
import { translateToRequest } from './transformers/request';
import { translateToResponses } from './transformers/responses';
import { translateToSecurities } from './transformers/securities';
import { translateToServers } from './transformers/servers';

export function transformOas2Operations<T extends DeepPartial<Spec> = DeepPartial<Spec>>(
  document: T,
  ctx?: TransformerContext<T>,
): IHttpOperation[] {
  return transformOasEndpointOperations(document, transformOas2Operation, OPERATION_CONFIG, void 0, ctx);
}

export const transformOas2Operation: Oas2HttpOperationTransformer = ({
  document: _document,
  name,
  method,
  config,
  ctx = createContext(_document),
}) => {
  const httpEndpointOperation = transformOasEndpointOperation.call(ctx, config, name, method);
  const parentObj = ctx.maybeResolveLocalRef(ctx.document[config.documentProp]![name]) as Fragment;
  const obj = ctx.maybeResolveLocalRef(parentObj[method]) as Fragment;

  return {
    ...httpEndpointOperation,

    responses: translateToResponses.call(ctx, obj),
    servers: translateToServers.call(ctx, obj),
    request: translateToRequest.call(ctx, parentObj, obj),
    security: translateToSecurities.call(ctx, obj.security, 'requirement'),
  } as any;
};
