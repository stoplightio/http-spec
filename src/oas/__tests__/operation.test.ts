import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';

import { transformOas2Operations } from '../../oas2/operation';
import { transformOas3Operations } from '../../oas3/operation';

const oas2KitchenSinkJson = JSON.parse(
  await fs.promises.readFile(path.join(fileURLToPath(import.meta.url), '../fixtures/oas2-kitchen-sink.json'), 'utf8'),
);
const oas3KitchenSinkJson = JSON.parse(
  await fs.promises.readFile(path.join(fileURLToPath(import.meta.url), '../fixtures/oas3-kitchen-sink.json'), 'utf8'),
);

describe('oas operation', () => {
  it('openapi v2', () => {
    const result = transformOas2Operations(oas2KitchenSinkJson);

    expect(result).toHaveLength(5);
    expect(result).toMatchSnapshot();
  });

  it('openapi v3', () => {
    const result = transformOas3Operations(oas3KitchenSinkJson);

    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });
});
