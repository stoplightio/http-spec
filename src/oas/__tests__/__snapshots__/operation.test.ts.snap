// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oas operation openapi v2 1`] = `
Array [
  Object {
    "description": "",
    "extensions": Object {
      "x-another": Object {
        "property": "value",
      },
      "x-test": true,
    },
    "id": "9ab1c1339675f",
    "iid": "addPet",
    "method": "post",
    "path": "/pets",
    "request": Object {
      "body": Object {
        "contents": Array [
          Object {
            "examples": Array [],
            "id": "beb02d1ff7b23",
            "mediaType": "application/json",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/definitions/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
          Object {
            "examples": Array [],
            "id": "531dc73646a3e",
            "mediaType": "application/xml",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/definitions/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
        ],
        "description": "Pet object that needs to be added to the store",
        "id": "7de2e97c373c6",
        "required": true,
      },
      "cookie": Array [],
      "headers": Array [],
      "path": Array [],
      "query": Array [],
    },
    "responses": Array [
      Object {
        "code": "405",
        "contents": Array [],
        "description": "Invalid input",
        "headers": Array [],
        "id": "2e992aee1674b",
      },
    ],
    "security": Array [
      Array [
        Object {
          "flows": Object {
            "implicit": Object {
              "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
              "scopes": Object {
                "read:pets": "read your pets",
                "write:pets": "modify pets in your account",
              },
            },
          },
          "id": "7d917a06422fe",
          "key": "petstore_auth",
          "type": "oauth2",
        },
      ],
    ],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "Add a new pet to the store",
    "tags": Array [
      Object {
        "id": "bfc2f4aa9bced",
        "name": "pet",
      },
    ],
  },
  Object {
    "extensions": Object {},
    "id": "29e0f9745e980",
    "method": "get",
    "path": "/pets",
    "request": Object {
      "cookie": Array [],
      "headers": Array [
        Object {
          "id": "b93f4ef60cf1d",
          "name": "Rate-Limit",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "9360fa7dc92ab",
            },
          },
          "style": "simple",
        },
      ],
      "path": Array [],
      "query": Array [
        Object {
          "id": "e1c08deab8b36",
          "name": "skip",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "828c85d6a2307",
            },
          },
          "style": "form",
        },
        Object {
          "id": "d04bcf1c141cf",
          "name": "limit",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "cdfdf74babd0b",
            },
          },
          "style": "form",
        },
      ],
    },
    "responses": Array [
      Object {
        "code": "200",
        "contents": Array [],
        "description": "",
        "headers": Array [],
        "id": "75d497d0ca810",
      },
    ],
    "security": Array [],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "List pets",
    "tags": Array [],
  },
  Object {
    "extensions": Object {},
    "id": "8bd67834e0236",
    "method": "options",
    "path": "/pets",
    "request": Object {
      "cookie": Array [],
      "headers": Array [
        Object {
          "id": "b93f4ef60cf1d",
          "name": "Rate-Limit",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "9360fa7dc92ab",
            },
          },
          "style": "simple",
        },
      ],
      "path": Array [],
      "query": Array [
        Object {
          "id": "e1c08deab8b36",
          "name": "skip",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "828c85d6a2307",
            },
          },
          "style": "form",
        },
        Object {
          "id": "d04bcf1c141cf",
          "name": "limit",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "cdfdf74babd0b",
            },
          },
          "style": "form",
        },
      ],
    },
    "responses": Array [
      Object {
        "code": "200",
        "contents": Array [],
        "description": "",
        "headers": Array [
          Object {
            "description": "Allowed clients",
            "id": "b1cba0180c7df",
            "name": "Allow",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "type": "string",
              "x-stoplight": Object {
                "id": "96fc94cefd7d2",
              },
            },
            "style": "simple",
          },
          Object {
            "description": "Remaining requests",
            "id": "32d8a9b2021a7",
            "name": "X-Rate-Limit",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "format": "int64",
              "maximum": 9223372036854776000,
              "minimum": -9223372036854776000,
              "type": "number",
              "x-stoplight": Object {
                "id": "3c572d09f18de",
              },
            },
            "style": "simple",
          },
        ],
        "id": "ae2d0e59a4075",
      },
    ],
    "security": Array [],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "List pets",
    "tags": Array [],
  },
  Object {
    "description": "",
    "extensions": Object {},
    "id": "4f51735d06d77",
    "method": "delete",
    "path": "/pets",
    "request": Object {
      "cookie": Array [],
      "headers": Array [],
      "path": Array [],
      "query": Array [],
    },
    "responses": Array [
      Object {
        "code": "400",
        "contents": Array [],
        "description": "Invalid ID supplied",
        "headers": Array [],
        "id": "24183a40dad2f",
      },
      Object {
        "code": "404",
        "contents": Array [],
        "description": "Pet not found",
        "headers": Array [],
        "id": "4a183a402036c",
      },
    ],
    "security": Array [],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "Deletes a pet",
    "tags": Array [],
  },
  Object {
    "description": "",
    "extensions": Object {},
    "id": "fc497ea5adb47",
    "iid": "updatePet",
    "method": "put",
    "path": "/pet/{petId}",
    "request": Object {
      "body": Object {
        "contents": Array [
          Object {
            "examples": Array [],
            "id": "04c11541cf13f",
            "mediaType": "application/json",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/definitions/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
          Object {
            "examples": Array [],
            "id": "3be1a46e035af",
            "mediaType": "application/xml",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/definitions/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
        ],
        "description": "Pet object that needs to be added to the store",
        "id": "46cce04bfdb01",
        "required": true,
      },
      "cookie": Array [],
      "headers": Array [],
      "path": Array [
        Object {
          "id": "1ee79143656f4",
          "name": "petId",
          "required": true,
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "3c1327003ffa9",
            },
          },
          "style": "simple",
        },
      ],
      "query": Array [],
    },
    "responses": Array [
      Object {
        "code": "400",
        "contents": Array [],
        "description": "Invalid ID supplied",
        "headers": Array [],
        "id": "9161bca4d6ce5",
      },
      Object {
        "code": "404",
        "contents": Array [
          Object {
            "examples": Array [],
            "id": "e4c3ee9ac9603",
            "mediaType": "application/json",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "code": Object {
                  "type": "string",
                },
                "message": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "code",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "d1d03adc8237d",
              },
            },
          },
          Object {
            "examples": Array [],
            "id": "78ade5a6a574a",
            "mediaType": "application/xml",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "code": Object {
                  "type": "string",
                },
                "message": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "code",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "d1d03adc8237d",
              },
            },
          },
        ],
        "description": "Our shared 404 response.",
        "headers": Array [],
        "id": "53fa597d5f695",
      },
      Object {
        "code": "405",
        "contents": Array [],
        "description": "Validation exception",
        "headers": Array [],
        "id": "3b7b1912287ef",
      },
    ],
    "security": Array [
      Array [
        Object {
          "flows": Object {
            "implicit": Object {
              "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
              "scopes": Object {
                "read:pets": "read your pets",
                "write:pets": "modify pets in your account",
              },
            },
          },
          "id": "7d917a06422fe",
          "key": "petstore_auth",
          "type": "oauth2",
        },
      ],
    ],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "Update an existing pet",
    "tags": Array [
      Object {
        "id": "bfc2f4aa9bced",
        "name": "pet",
      },
    ],
  },
]
`;

exports[`oas operation openapi v3 1`] = `
Array [
  Object {
    "description": "",
    "extensions": Object {
      "x-another": Object {
        "property": "value",
      },
      "x-test": true,
    },
    "id": "9ab1c1339675f",
    "iid": "addPet",
    "method": "post",
    "path": "/pets",
    "request": Object {
      "body": Object {
        "contents": Array [
          Object {
            "encodings": Array [],
            "examples": Array [
              Object {
                "id": "97acd8b5018d8",
                "key": "Dog",
                "summary": "A dog example",
                "value": Object {
                  "name": "Leo",
                  "photoUrls": Array [],
                },
              },
              Object {
                "externalValue": "https://stoplight.io/pets/cat",
                "id": "97acd8b5018d8",
                "key": "Cat",
                "summary": "A cat example",
              },
            ],
            "id": "954d8d8652606",
            "mediaType": "application/json",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/components/schemas/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
          Object {
            "encodings": Array [],
            "examples": Array [],
            "id": "1043a7cddde88",
            "mediaType": "application/xml",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/components/schemas/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
        ],
        "description": "Pet object that needs to be added to the store",
        "id": "9b5d65cfbc4fd",
        "required": true,
      },
      "cookie": Array [],
      "headers": Array [],
      "path": Array [],
      "query": Array [],
    },
    "responses": Array [
      Object {
        "code": "405",
        "contents": Array [],
        "description": "Invalid input",
        "headers": Array [],
        "id": "029110671da4a",
      },
    ],
    "security": Array [
      Array [
        Object {
          "flows": Object {
            "implicit": Object {
              "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
              "scopes": Object {
                "read:pets": "read your pets",
                "write:pets": "modify pets in your account",
              },
            },
          },
          "id": "7d917a06422fe",
          "key": "petstore_auth",
          "type": "oauth2",
        },
      ],
    ],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "Add a new pet to the store",
    "tags": Array [
      Object {
        "id": "bfc2f4aa9bced",
        "name": "pet",
      },
    ],
  },
  Object {
    "extensions": Object {},
    "id": "29e0f9745e980",
    "method": "get",
    "path": "/pets",
    "request": Object {
      "cookie": Array [],
      "headers": Array [],
      "path": Array [],
      "query": Array [
        Object {
          "examples": Array [],
          "id": "e1c08deab8b36",
          "name": "skip",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "828c85d6a2307",
            },
          },
          "style": "form",
        },
        Object {
          "examples": Array [],
          "id": "d04bcf1c141cf",
          "name": "limit",
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "cdfdf74babd0b",
            },
          },
          "style": "form",
        },
      ],
    },
    "responses": Array [
      Object {
        "code": "200",
        "contents": Array [
          Object {
            "encodings": Array [],
            "examples": Array [],
            "id": "61b0e53aa1cce",
            "mediaType": "*/*",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {},
              "type": "object",
              "x-stoplight": Object {
                "id": "8ea4aec4c2854",
              },
            },
          },
        ],
        "description": "OK",
        "headers": Array [],
        "id": "6ae68ec4738b2",
      },
    ],
    "security": Array [],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "List pets",
    "tags": Array [],
  },
  Object {
    "description": "",
    "extensions": Object {},
    "id": "fc497ea5adb47",
    "iid": "updatePet",
    "method": "put",
    "path": "/pet/{petId}",
    "request": Object {
      "body": Object {
        "contents": Array [
          Object {
            "encodings": Array [],
            "examples": Array [
              Object {
                "id": "97acd8b5018d8",
                "key": "Dog",
                "summary": "A dog example",
                "value": Object {
                  "name": "Leo",
                  "photoUrls": Array [],
                },
              },
              Object {
                "externalValue": "https://stoplight.io/pets/cat",
                "id": "97acd8b5018d8",
                "key": "Cat",
                "summary": "A cat example",
              },
            ],
            "id": "954d8d8652606",
            "mediaType": "application/json",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/components/schemas/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
          Object {
            "encodings": Array [],
            "examples": Array [],
            "id": "1043a7cddde88",
            "mediaType": "application/xml",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "category": Object {
                  "$ref": "#/components/schemas/Category",
                },
                "id": Object {
                  "format": "int64",
                  "maximum": 9223372036854776000,
                  "minimum": -9223372036854776000,
                  "type": "integer",
                },
                "name": Object {
                  "examples": Array [
                    "doggie",
                  ],
                  "type": "string",
                },
                "photoUrls": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                  "xml": Object {
                    "name": "photoUrl",
                    "wrapped": true,
                  },
                },
                "status": Object {
                  "description": "pet status in the store",
                  "enum": Array [
                    "available",
                    "pending",
                    "sold",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "photoUrls",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "2b4cfd51fbcde",
              },
              "xml": Object {
                "name": "Pet",
              },
            },
          },
        ],
        "description": "Pet object that needs to be added to the store",
        "id": "9b5d65cfbc4fd",
        "required": true,
      },
      "cookie": Array [],
      "headers": Array [],
      "path": Array [
        Object {
          "examples": Array [],
          "id": "1ee79143656f4",
          "name": "petId",
          "required": true,
          "schema": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "type": "string",
            "x-stoplight": Object {
              "id": "3c1327003ffa9",
            },
          },
          "style": "simple",
        },
      ],
      "query": Array [],
    },
    "responses": Array [
      Object {
        "code": "400",
        "contents": Array [],
        "description": "Invalid ID supplied",
        "headers": Array [],
        "id": "c12ebf63598e3",
      },
      Object {
        "code": "404",
        "contents": Array [
          Object {
            "encodings": Array [],
            "examples": Array [],
            "id": "edbd8b027c519",
            "mediaType": "*/*",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "code": Object {
                  "type": "string",
                },
                "message": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "code",
              ],
              "type": "object",
              "x-stoplight": Object {
                "id": "d1d03adc8237d",
              },
            },
          },
        ],
        "description": "Our shared 404 response.",
        "headers": Array [],
        "id": "e1deb92ee1e73",
      },
      Object {
        "code": "405",
        "contents": Array [],
        "description": "Validation exception",
        "headers": Array [],
        "id": "c15ebf6359d8c",
      },
    ],
    "security": Array [
      Array [
        Object {
          "flows": Object {
            "implicit": Object {
              "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
              "scopes": Object {
                "read:pets": "read your pets",
                "write:pets": "modify pets in your account",
              },
            },
          },
          "id": "7d917a06422fe",
          "key": "petstore_auth",
          "type": "oauth2",
        },
      ],
    ],
    "servers": Array [
      Object {
        "id": "be714555a4b1c",
        "name": "Swagger Petstore",
        "url": "https://petstore.swagger.io/v2",
      },
      Object {
        "id": "b9d12ab472c47",
        "name": "Swagger Petstore",
        "url": "http://petstore.swagger.io/v2",
      },
    ],
    "summary": "Update an existing pet",
    "tags": Array [
      Object {
        "id": "bfc2f4aa9bced",
        "name": "pet",
      },
    ],
  },
]
`;
